### PR TITLE
REPL: Add variables view

### PR DESCRIPTION
**Pull request by dima74**
Add variables view to REPL console. It shows available variables, functions and structs (and their types)
Screen Shot 2020-03-16 at 11 56 37

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/presentation/Utils.kt [hunk 1]
```diff
@@ -76,6 +90,7 @@ private fun presentableName(psi: RsElement): String? {
                 append(typeParameterBounds(psi))
             }
         }
+        is RsLetDecl -> (psi.pat as? RsPatIdentImpl)?.patBinding?.name
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/presentation/Utils.kt]: Do you really need `RsPatIdentImpl` here and below (`RsStructureViewModel.kt`), or maybe just `RsPatIdent`?

Comment (ortem) [inline src/main/kotlin/org/rust/ide/presentation/Utils.kt]: And a more interesting problem: your implementation doesn't handle let declarations with complex patterns. E.g. `let (a,b) = (1,2)` will be ignored in the variables view. But I think it should not. At least, Python Console supports tuple patterns in declarations. Probably you should collect all the bindings like `val bindings = letDecl.pat.descendantsOfType<RsPatBinding>()` and then handle each binding separately.

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsStmt.kt [hunk 2]
```diff
@@ -8,8 +8,21 @@ package org.rust.lang.core.psi.ext
...
+import org.rust.lang.core.psi.RsLetDecl
+import org.rust.lang.core.psi.RsPatIdent
 import org.rust.lang.core.psi.RsStmt
+import javax.swing.Icon
 
 abstract class RsStmtMixin(node: ASTNode) : RsElementImpl(node), RsStmt {
     override fun getContext(): PsiElement? = RsExpandedElement.getContextImpl(this)
+
+    override fun getIcon(flags: Int): Icon? {
+        if (this is RsLetDecl) {
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsStmt.kt]: Please add a comment that this is used only for REPL code fragments

File: src/main/kotlin/org/rust/ide/console/RsConsoleVariablesView.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    init {
+        val allCommands = codeFragmentContext.getAllCommandsText()
+        variablesFile = PsiFileFactory.getInstance(project)
+            .createFileFromText(RsConsoleView.VIRTUAL_FILE_NAME, RsLanguage, allCommands) as RsReplCodeFragment
+        val structureViewModel = RsStructureViewModel(null, variablesFile)
+        treeStructure = SmartTreeStructure(project, structureViewModel)
+
+        structureTreeModel = StructureTreeModel(treeStructure, this)
+        val asyncTreeModel = AsyncTreeModel(structureTreeModel, this)
+        asyncTreeModel.setRootImmediately(structureTreeModel.rootImmediately)
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/console/RsConsoleVariablesView.kt]: This method is deprecated. Do we really need it?

File: src/main/kotlin/org/rust/ide/console/RsConsoleVariablesView.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        val tree = Tree(asyncTreeModel)
+        tree.isRootVisible = false
+        tree.emptyText.text = EMPTY_TEXT
+
+        setContent(ScrollPaneFactory.createScrollPane(tree))
+    }
+
+    fun rebuild() {
+        GuiUtils.invokeLaterIfNeeded({
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/console/RsConsoleVariablesView.kt]: Please use `runInEdt` instead (https://github.com/intellij-rust/intellij-rust/pull/4918)

File: src/main/kotlin/org/rust/ide/presentation/Utils.kt (hunk 1 above)
Comment (dima74) [reply src/main/kotlin/org/rust/ide/presentation/Utils.kt]: Yes, `RsPatIdent` is enough

Comment (dima74) [reply src/main/kotlin/org/rust/ide/presentation/Utils.kt]: Yes, fixed. Now all variables (`RsPatIdent`) are collected in `RsStructureViewModel::getVariablesDeclarations`

File: src/main/kotlin/org/rust/ide/console/RsConsoleVariablesView.kt (hunk 3 above)
Comment (dima74) [reply src/main/kotlin/org/rust/ide/console/RsConsoleVariablesView.kt]: I tested that tree model works without this method, so removed

**Comment by dima74** [comment]
@ortem Thank you for comments, I updated code and rebased

File: src/main/kotlin/org/rust/ide/console/RsConsoleView.kt [hunk 5]
```diff
@@ -73,9 +78,54 @@ class RsConsoleView(project: Project) : LanguageConsoleImpl(project, VIRTUAL_FIL
         if (codeFragment != null) {
             codeFragmentContext.addToContext(lastCommandContext)
             codeFragmentContext.updateContext(project, codeFragment)
+            variablesView?.rebuild()
         }
     }
 
+    private fun splitWindow() {
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/console/RsConsoleView.kt]: For now, the logic is not obvious IMO. Maybe get rid of `showVariables` and do everything in-place?  
```Kotlin
private fun showVariables() {
    val console = getComponent(0)
    if (variablesView == null) {
        variablesView = RsConsoleVariablesView(project, codeFragmentContext)
        removeAll()
        // ...
    }
}
```

File: src/main/kotlin/org/rust/ide/console/RsConsoleView.kt [hunk 6]
```diff
@@ -73,9 +78,54 @@ class RsConsoleView(project: Project) : LanguageConsoleImpl(project, VIRTUAL_FIL
...
+            secondComponent = variablesView as JComponent
+            isShowDividerControls = true
+            setHonorComponentsMinimumSize(true)
+        }
+        add(splitter, BorderLayout.CENTER)
+        validate()
+        repaint()
+    }
+
+    fun restoreWindow() {
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/console/RsConsoleView.kt]:
```suggestion
    private fun hideVariables() {
```

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsElement.kt [hunk 7]
```diff
@@ -93,6 +91,14 @@ abstract class RsElementImpl(node: ASTNode) : ASTWrapperPsiElement(node), RsElem
 
     final override val crateRoot: RsMod?
         get() = (contextualFile as? RsElement)?.crateRoot
+
+    // used only for rust console (RsReplCodeFragment)
+    override fun getIcon(flags: Int): Icon? {
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsElement.kt]: Also, maybe put this logic into `RsPatBindingImplMixin` or something related to `PatBinding`, but not `PatIdent`?

File: src/main/kotlin/org/rust/ide/structure/RsStructureViewModel.kt [hunk 8]
```diff
@@ -96,3 +97,14 @@ private class RsStructureViewElement(
             }
         }
 }
+
+private fun RsReplCodeFragment.getVariablesDeclarations(): Collection<RsPatIdent> {
+    val variables = mutableMapOf<String, RsPatIdent>()
+    for (letDecl in childrenOfType<RsLetDecl>()) {
+        for (pat in letDecl.descendantsOfType<RsPatIdent>()) {
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/ide/structure/RsStructureViewModel.kt]: Yes, `RsPatBinding` is better. For some reason I didn't notice that you suggest to use it in first review

File: src/main/kotlin/org/rust/ide/console/RsConsoleActions.kt [hunk 9]
```diff
@@ -101,3 +101,21 @@ class PrintAction(private val consoleView: RsConsoleView) : DumbAwareAction() {
...
 }
+
+class ShowVarsAction(private val consoleView: RsConsoleView)
+    : ToggleAction("Show Variables", "Shows active console variables", AllIcons.Debugger.Watch), DumbAware {
+
+    override fun isSelected(e: AnActionEvent): Boolean {
+        return consoleView.options.isShowVars
+    }
+
+    override fun setSelected(e: AnActionEvent, state: Boolean) {
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/ide/console/RsConsoleActions.kt]: Yes, good idea

File: src/main/kotlin/org/rust/ide/console/RsConsoleVariablesView.kt [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    fun rebuild() {
+        runInEdt {
+            runWriteAction {
+                val allCommands = codeFragmentContext.getAllCommandsText()
+                variablesFile.virtualFile.document?.setText(allCommands)
+            }
+
+            structureTreeModel.invoker.runOrInvokeLater {
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/ide/console/RsConsoleVariablesView.kt]: Yes, this code doesn't need EDT, but it should run after `runWriteAction` block. I replaced `runOrInvokeLater` with `invokeLater` to be sure that code doesn't run on EDT (currently `runOrInvokeLater` already acts as `invokeLater`)

**Comment by dima74** [comment]
@Undin yes, I added some tests

File: src/main/kotlin/org/rust/ide/presentation/Utils.kt [hunk 11]
```diff
@@ -76,6 +80,7 @@ private fun presentableName(psi: RsElement): String? {
                 append(typeParameterBounds(psi))
             }
         }
+        is RsPatBinding -> psi.name
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/presentation/Utils.kt]: `RsPatBinding` is `RsNamedElement` which is processed above

**Comment by dima74** [comment]
bors r=ortem

**Comment by Undin** [comment]
bors retry

**Pull request merged**